pkg/helm/charts.go

```diff
@@ -0,0 +1,283 @@
+	}
+
+	chart, err := h.loadChart(ctx, helmRepo, c)
+	if err != nil {
+		return nil, fmt.Errorf("loading %s from chart: %w", filename, err)
+	}
+
+	for _, v := range chart.Raw {
+		if v.Name == filename {
+			return v.Data, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find file: %s", filename)
+}
+
+func (h *RepoManager) updateCache(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository) error {
+	entry, err := h.entryForRepository(ctx, helmRepo)
+	if err != nil {
+		return err
+	}
+
+	r, err := repo.NewChartRepository(entry, DefaultChartGetters)
+	if err != nil {
+		return fmt.Errorf("error creating chart repository: %w", err)
+	}
+
+	r.CachePath = h.CacheDir
+	_, err = r.DownloadIndexFile()
+
+	return err
+}
+
+func (h *RepoManager) loadChart(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository, c *ChartReference) (*chart.Chart, error) {
+	o, err := h.chartPathOptionsFromRepository(ctx, helmRepo, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to configure client: %w", err)
+	}
+
+	chartLocation, err := o.LocateChart(c.Chart, h.envSettings)
+	if err != nil {
+		return nil, fmt.Errorf("locating chart %q: %w", c.Chart, err)
+	}
+
+	chart, err := loader.Load(chartLocation)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load chart %q: %w", c.Chart, err)
+	}
+
+	return chart, nil
+}
+
+func (h *RepoManager) chartPathOptionsFromRepository(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository, c *ChartReference) (*action.ChartPathOptions, error) {
+	// TODO: This should probably use Verify: true
+	co := &action.ChartPathOptions{
+		RepoURL: helmRepo.Spec.URL,
+		Version: c.Version,
+	}
+
+	if helmRepo.Spec.SecretRef != nil {
+		username, password, err := credsForRepository(ctx, h.Client, helmRepo)
+		if err != nil {
+			return nil, err
+		}
+
+		co.Username = username
+		co.Password = password
+	}
+
+	return co, nil
+}
+
+func (h *RepoManager) entryForRepository(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository) (*repo.Entry, error) {
+	entry := &repo.Entry{
+		Name: helmRepo.GetName() + "-" + helmRepo.GetNamespace(),
+		URL:  helmRepo.Spec.URL,
+	}
+
+	if helmRepo.Spec.SecretRef != nil {
+		username, password, err := credsForRepository(ctx, h.Client, helmRepo)
+		if err != nil {
+			return nil, err
+		}
+
+		entry.Username = username
+		entry.Password = password
+	}
+
+	return entry, nil
+}
+
+func credsForRepository(ctx context.Context, kc client.Client, hr *sourcev1beta1.HelmRepository) (string, string, error) {
+	var secret corev1.Secret
+	if err := kc.Get(ctx, types.NamespacedName{Name: hr.Spec.SecretRef.Name, Namespace: hr.Namespace}, &secret); err != nil {
+		return "", "", fmt.Errorf("repository authentication: %w", err)
+	}
+
+	return string(secret.Data["username"]), string(secret.Data["password"]), nil
+}
+
+func fetchIndexFile(chartURL string) (*repo.IndexFile, error) {
+	if hostname := os.Getenv("SOURCE_CONTROLLER_LOCALHOST"); hostname != "" {
```

Comment:
What is the purpose of this environment property? I'm only seeing it here and not anywhere other and there is no documentation of it.

Reply:
If this is for testing purposes only to change the hostname to a local test service, we should either hide it better or make it injectable.

Reply:
removing it since we aren't using it